moshi/src/main/java/com/squareup/moshi/BufferedSourceJsonReader.java

```diff
@@ -236,6 +237,11 @@
   }
 
   private int doPeek() throws IOException {
+    if (stackSize == 1
```

Comment:
I'm thinking this should be below `peekStack` and check for `EMPTY_DOCUMENT` instead? Otherwise we potentially match the BOM anywhere in the root object.